File: src/lib/simpleBinaryProtocol.ts
// シンプルなバイナリプロトコルの実装
// 軽量で高速な描画コマンドのエンコード/デコード

import type { DrawEngineCommand } from './bindings';

export class SimpleBinaryProtocol {
  // コマンドタイプの定数
  private static readonly CMD_BEGIN_STROKE = 0;
  private static readonly CMD_CONTINUE_STROKE = 1;
  private static readonly CMD_END_STROKE = 2;
  private static readonly CMD_CLEAR = 3;
  private static readonly CMD_SET_BRUSH = 4;
  private static readonly CMD_SET_ACTIVE_LAYER = 5;
  private static readonly CMD_CREATE_LAYER = 6;
  private static readonly CMD_DELETE_LAYER = 7;

  // 単一のDrawEngineCommandをバイナリ形式にエンコード
  static encodeCommand(command: DrawEngineCommand): Uint8Array {
    // beginStroke
    if (typeof command === 'object' && 'beginStroke' in command) {
      const { x, y, pressure } = command.beginStroke;
      const buffer = new ArrayBuffer(13); // 1 (cmd) + 4*3 (floats)
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_BEGIN_STROKE);
      view.setFloat32(1, x, true);
      view.setFloat32(5, y, true);
      view.setFloat32(9, pressure, true);
      return new Uint8Array(buffer);
    }
    
    // continueStroke
    if (typeof command === 'object' && 'continueStroke' in command) {
      const { x, y, pressure } = command.continueStroke;
      const buffer = new ArrayBuffer(13); // 1 (cmd) + 4*3 (floats)
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_CONTINUE_STROKE);
      view.setFloat32(1, x, true);
      view.setFloat32(5, y, true);
      view.setFloat32(9, pressure, true);
      return new Uint8Array(buffer);
    }
    
    // endStroke
    if (command === 'endStroke') {
      const buffer = new ArrayBuffer(1);
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_END_STROKE);
      return new Uint8Array(buffer);
    }
    
    // clear
    if (command === 'clear') {
      const buffer = new ArrayBuffer(1);
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_CLEAR);
      return new Uint8Array(buffer);
    }
    
    // setBrush
    if (typeof command === 'object' && 'setBrush' in command) {
      const brush = command.setBrush;
      const buffer = new ArrayBuffer(33); // 1 (cmd) + 32 (brush data)
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_SET_BRUSH);
      
      let offset = 1;
      view.setFloat32(offset, brush.size, true);
      offset += 4;
      view.setFloat32(offset, brush.opacity, true);
      offset += 4;
      
      // color: [f32; 4]
      for (let i = 0; i < 4; i++) {
        view.setFloat32(offset, brush.color[i], true);
        offset += 4;
      }
      
      // brushType (as u32)
      const brushTypeValue = brush.brushType === 'pen' ? 0 : brush.brushType === 'brush' ? 1 : 2;
      view.setUint32(offset, brushTypeValue, true);
      offset += 4;
      
      // blendMode (as u32)
      const blendModeValue = 
        brush.blendMode === 'normal' ? 0 :
        brush.blendMode === 'multiply' ? 1 :
        brush.blendMode === 'screen' ? 2 : 3;
      view.setUint32(offset, blendModeValue, true);
      
      return new Uint8Array(buffer);
    }
    
    // setActiveLayer
    if (typeof command === 'object' && 'setActiveLayer' in command) {
      const layerIndex = command.setActiveLayer;
      const buffer = new ArrayBuffer(9); // 1 (cmd) + 8 (u64)
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_SET_ACTIVE_LAYER);
      view.setBigUint64(1, BigInt(layerIndex), true);
      return new Uint8Array(buffer);
    }
    
    // createLayer
    if (command === 'createLayer') {
      const buffer = new ArrayBuffer(1);
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_CREATE_LAYER);
      return new Uint8Array(buffer);
    }
    
    // deleteLayer
    if (typeof command === 'object' && 'deleteLayer' in command) {
      const layerIndex = command.deleteLayer;
      const buffer = new ArrayBuffer(9); // 1 (cmd) + 8 (u64)
      const view = new DataView(buffer);
      view.setUint8(0, this.CMD_DELETE_LAYER);
      view.setBigUint64(1, BigInt(layerIndex), true);
      return new Uint8Array(buffer);
    }
    
    throw new Error(`Unknown command type: ${JSON.stringify(command)}`);
  }
  
  // 複数のコマンドをバッチとしてエンコード
  static encodeCommandBatch(commands: DrawEngineCommand[]): Uint8Array {
    // まず各コマンドをエンコード
    const encodedCommands = commands.map(cmd => this.encodeCommand(cmd));
    
    // 全体のサイズを計算
    const totalSize = encodedCommands.reduce((sum, cmd) => sum + cmd.length + 4, 4);
    // 4 bytes for command count + (4 bytes length + data) per command
    
    const buffer = new ArrayBuffer(totalSize);
    const view = new DataView(buffer);
    const uint8View = new Uint8Array(buffer);
    
    let offset = 0;
    
    // コマンド数を書き込み
    view.setUint32(offset, commands.length, true);
    offset += 4;
    
    // 各コマンドを書き込み
    for (const cmdData of encodedCommands) {
      // コマンドの長さ
      view.setUint32(offset, cmdData.length, true);
      offset += 4;
      
      // コマンドデータ
      uint8View.set(cmdData, offset);
      offset += cmdData.length;
    }
    
    return uint8View;
  }
}

File: src-tauri/src/ipc/simple_binary_protocol.rs
use crate::drawing_engine::commands::{DrawEngineCommand, BrushSettings, BrushType, BlendMode};
use std::io::Cursor;
use byteorder::{LittleEndian, ReadBytesExt};

pub struct SimpleBinaryProtocol;

impl SimpleBinaryProtocol {
    // コマンドタイプの定数
    const CMD_BEGIN_STROKE: u8 = 0;
    const CMD_CONTINUE_STROKE: u8 = 1;
    const CMD_END_STROKE: u8 = 2;
    const CMD_CLEAR: u8 = 3;
    const CMD_SET_BRUSH: u8 = 4;
    const CMD_SET_ACTIVE_LAYER: u8 = 5;
    const CMD_CREATE_LAYER: u8 = 6;
    const CMD_DELETE_LAYER: u8 = 7;
    
    /// 単一のコマンドをデコード
    pub fn decode_command(data: &[u8]) -> Result<DrawEngineCommand, Box<dyn std::error::Error>> {
        if data.is_empty() {
            return Err("Empty data".into());
        }
        
        let mut cursor = Cursor::new(data);
        let cmd_type = cursor.read_u8()?;
        
        match cmd_type {
            Self::CMD_BEGIN_STROKE => {
                let x = cursor.read_f32::<LittleEndian>()?;
                let y = cursor.read_f32::<LittleEndian>()?;
                let pressure = cursor.read_f32::<LittleEndian>()?;
                Ok(DrawEngineCommand::BeginStroke { x, y, pressure })
            }
            
            Self::CMD_CONTINUE_STROKE => {
                let x = cursor.read_f32::<LittleEndian>()?;
                let y = cursor.read_f32::<LittleEndian>()?;
                let pressure = cursor.read_f32::<LittleEndian>()?;
                Ok(DrawEngineCommand::ContinueStroke { x, y, pressure })
            }
            
            Self::CMD_END_STROKE => {
                Ok(DrawEngineCommand::EndStroke)
            }
            
            Self::CMD_CLEAR => {
                Ok(DrawEngineCommand::Clear)
            }
            
            Self::CMD_SET_BRUSH => {
                let size = cursor.read_f32::<LittleEndian>()?;
                let opacity = cursor.read_f32::<LittleEndian>()?;
                
                let mut color = [0.0; 4];
                for i in 0..4 {
                    color[i] = cursor.read_f32::<LittleEndian>()?;
                }
                
                let brush_type_val = cursor.read_u32::<LittleEndian>()?;
                let brush_type = match brush_type_val {
                    0 => BrushType::Pen,
                    1 => BrushType::Brush,
                    2 => BrushType::Eraser,
                    _ => return Err(format!("Invalid brush type: {}", brush_type_val).into()),
                };
                
                let blend_mode_val = cursor.read_u32::<LittleEndian>()?;
                let blend_mode = match blend_mode_val {
                    0 => BlendMode::Normal,
                    1 => BlendMode::Multiply,
                    2 => BlendMode::Screen,
                    3 => BlendMode::Overlay,
                    _ => return Err(format!("Invalid blend mode: {}", blend_mode_val).into()),
                };
                
                Ok(DrawEngineCommand::SetBrush(BrushSettings {
                    size,
                    opacity,
                    color,
                    brush_type,
                    blend_mode,
                }))
            }
            
            Self::CMD_SET_ACTIVE_LAYER => {
                let layer_index = cursor.read_u64::<LittleEndian>()? as usize;
                Ok(DrawEngineCommand::SetActiveLayer(layer_index))
            }
            
            Self::CMD_CREATE_LAYER => {
                Ok(DrawEngineCommand::CreateLayer)
            }
            
            Self::CMD_DELETE_LAYER => {
                let layer_index = cursor.read_u64::<LittleEndian>()? as usize;
                Ok(DrawEngineCommand::DeleteLayer(layer_index))
            }
            
            _ => Err(format!("Unknown command type: {}", cmd_type).into()),
        }
    }
    
    /// バッチコマンドをデコード
    pub fn decode_commands(data: &[u8]) -> Result<Vec<DrawEngineCommand>, Box<dyn std::error::Error>> {
        let mut cursor = Cursor::new(data);
        let command_count = cursor.read_u32::<LittleEndian>()? as usize;
        
        let mut commands = Vec::with_capacity(command_count);
        
        for _ in 0..command_count {
            let cmd_length = cursor.read_u32::<LittleEndian>()? as usize;
            
            // 現在の位置を取得
            let current_pos = cursor.position() as usize;
            
            // コマンドデータを取得
            if current_pos + cmd_length > data.len() {
                return Err("Insufficient data for command".into());
            }
            
            let cmd_data = &data[current_pos..current_pos + cmd_length];
            let command = Self::decode_command(cmd_data)?;
            commands.push(command);
            
            // カーソルを進める
            cursor.set_position((current_pos + cmd_length) as u64);
        }
        
        Ok(commands)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_decode_begin_stroke() {
        let data = vec![0, 0, 0, 128, 63, 0, 0, 0, 64, 0, 0, 0, 63]; // cmd=0, x=1.0, y=2.0, pressure=0.5
        let command = SimpleBinaryProtocol::decode_command(&data).unwrap();
        
        match command {
            DrawEngineCommand::BeginStroke { x, y, pressure } => {
                assert_eq!(x, 1.0);
                assert_eq!(y, 2.0);
                assert_eq!(pressure, 0.5);
            }
            _ => panic!("Wrong command type"),
        }
    }
    
    #[test]
    fn test_decode_commands_batch() {
        // バッチ: 2コマンド
        let mut data = vec![];
        
        // コマンド数 (2)
        data.extend_from_slice(&2u32.to_le_bytes());
        
        // コマンド1: beginStroke
        let cmd1 = vec![0, 0, 0, 128, 63, 0, 0, 0, 64, 0, 0, 0, 63]; // cmd=0, x=1.0, y=2.0, pressure=0.5
        data.extend_from_slice(&(cmd1.len() as u32).to_le_bytes());
        data.extend_from_slice(&cmd1);
        
        // コマンド2: endStroke
        let cmd2 = vec![2]; // cmd=2
        data.extend_from_slice(&(cmd2.len() as u32).to_le_bytes());
        data.extend_from_slice(&cmd2);
        
        let commands = SimpleBinaryProtocol::decode_commands(&data).unwrap();
        assert_eq!(commands.len(), 2);
        
        match &commands[0] {
            DrawEngineCommand::BeginStroke { x, y, pressure } => {
                assert_eq!(*x, 1.0);
                assert_eq!(*y, 2.0);
                assert_eq!(*pressure, 0.5);
            }
            _ => panic!("Wrong command type for first command"),
        }
        
        match &commands[1] {
            DrawEngineCommand::EndStroke => {},
            _ => panic!("Wrong command type for second command"),
        }
    }
}